pilot/pkg/xds/ads.go

```diff
@@ -601,32 +602,49 @@ func (s *DiscoveryServer) initProxyMetadata(node *core.Node) (*model.Proxy, erro
 // setTopologyLabels sets locality, cluster, network label
 // must be called after `SetWorkloadLabels` and `SetServiceInstances`.
 func setTopologyLabels(proxy *model.Proxy) {
-	var localityStr string
-	// Get the locality from the proxy's service instances.
-	// We expect all instances to have the same IP and therefore the same locality.
-	// So its enough to look at the first instance.
-	if len(proxy.ServiceInstances) > 0 {
-		localityStr = proxy.ServiceInstances[0].Endpoint.Locality.Label
-	} else {
-		// If no service instances(this maybe common for a pure client), respect LocalityLabel
-		localityStr = proxy.Labels[model.LocalityLabel]
+	// This is a bit un-intuitive, but pull the locality from Labels first. The service registries have the best access to
+	// locality information, as they can read from various sources (Node on Kubernetes, for example). They will take this
+	// information and add it to the labels (via AugmentLabels). So while the proxy may not originally have these labels,
+	// it will by the time we get here (as a result of calling this after SetWorkloadLabels). Note: this may also be from a
+	// previous iteration of setTopologyLabels on the proxy. This is fine; if the service registry has updated information
+	// it would override the labels, else we use the previously known value.
+	locality := localityFromProxyLabels(proxy)
+	if locality == nil {
+		// If no labels set, we didn't find the locality from the service registry. We do support a (mostly undocumented/internal)
+		// label to override the locality, so respect that here as well.
+		ls, f := proxy.Labels[model.LocalityLabel]
+		if f {
+			locality = util.ConvertLocality(ls)
+		}
 	}
-	if localityStr != "" {
-		proxy.Locality = util.ConvertLocality(localityStr)
-	} else {
+	if locality == nil {
 		// If there is no locality in the registry then use the one sent as part of the discovery request.
 		// This is not preferable as only the connected Pilot is aware of this proxies location, but it
 		// can still help provide some client-side Envoy context when load balancing based on location.
-		proxy.Locality = &core.Locality{
+		locality = &core.Locality{
 			Region:  proxy.XdsNode.Locality.GetRegion(),
 			Zone:    proxy.XdsNode.Locality.GetZone(),
 			SubZone: proxy.XdsNode.Locality.GetSubZone(),
 		}
 	}
-
-	locality := util.LocalityToString(proxy.Locality)
+	proxy.Locality = locality
 	// add topology labels to proxy labels
-	proxy.Labels = labelutil.AugmentLabels(proxy.Labels, proxy.Metadata.ClusterID, locality, proxy.GetNodeName(), proxy.Metadata.Network)
+	proxy.Labels = labelutil.AugmentLabels(proxy.Labels, proxy.Metadata.ClusterID, util.LocalityToString(proxy.Locality), proxy.GetNodeName(), proxy.Metadata.Network)
+}
+
+func localityFromProxyLabels(proxy *model.Proxy) *core.Locality {
+	region, f1 := proxy.Labels[labelutil.LabelTopologyRegion]
+	zone, f2 := proxy.Labels[labelutil.LabelTopologyZone]
+	subzone, f3 := proxy.Labels[label.TopologySubzone.Name]
```

Comment:
This seems not right as we donot set node labeles to proxy, we only set `istio-locality`. 

So maybe we get locality from Lebels[istio-locality] is enough

Reply:
we do from AugmentLabels?

Reply:
Yes, but AugmentLabels is called after calling this function.

Previouly we also call AugmentLabels in endpoint builder